Block/Catalog/Product/Usage.php

```diff
@@ -490,4 +519,166 @@ public function getCustomLicenseId()
+     */
+    public function getPreviousSelectHtml()
+    {
+        $extraParams = '';
+        $select = $this->getLayout()->createBlock(
+            \Magento\Framework\View\Element\Html\Select::class
+        )->setData(
+            [
+                'id' => 'usage_previous_usages',
+                'class' => 'required product-custom-option admin__control-select usage-select-box'
+            ]
+        );
+
+        $select->setName('previous_category')->addOption('', __('-- Please Select --'));
+
+        foreach ($this->getPreviousCategories() as $category) {
+
+            $select->addOption(
+                $category['id'],
+                $category['name']
+            );
+        }
+        $extraParams .= ' data-selector="' . $select->getName() . '"';
+        $select->setExtraParams($extraParams);
+
+        return $select->getHtml();
+    }
+
+    /**
+     * @return array
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function getPreviousCategories()
+    {
+        $previousCategories = [];
+        $items = $this->checkoutSession->getQuote()->getAllVisibleItems();
+
+        foreach ($items as $item) {
+            /**
+             * @var \Magento\Quote\Model\Quote\Item $item
+             */
+            $id = '';
+            $value = '';
+            $previousCategoriesByItems = [];
+            $getBuyRequest = $item->getBuyRequest();
+            foreach ($getBuyRequest as $data) {
+                if (isset($data['usage_category'])) {
+                    $id .= $data['usage_category'] . ' - ';
+                    $value .= $this->getCategoryName($data['usage_category']) . ' - ';
+                    $id .= $data['usage_id'][$data['usage_category']] . ' - ';
+                    $value .= $this->getUsageName($data['usage_id'][$data['usage_category']]) . ' - ';
+                    if (isset($data['options'])) {
+                        foreach ($data['options'] as $key => $option) {
+                            $id .= $key . ':' . $option . ' - ';
+                            $value .= $this->getOptionName($option) . ' - ';
+                        }
+                    }
+                }
+            }
+            if (!empty($id) && !empty($value)) {
+                $previousCategoriesByItems['id'] = rtrim(ltrim($id, ' - '), ' - ');
+                $previousCategoriesByItems['name'] = rtrim(ltrim($value, ' - '), ' - ');
+                if (!in_array($previousCategoriesByItems, $previousCategories)) {
+                    $previousCategories[] = $previousCategoriesByItems;
+                }
+            }
+            if (count($previousCategories) >= 10) {
+                break;
+            }
+        }
+        if ($this->isCustomerLoggedIn() && count($previousCategories) <= 10) {
+            $ordersCollection = $this->orderCollectionFactory->create()
```

Comment:
@superdav42 I also had the same concern about memory timeout with this approach but if you check the structure of orders in database, items are stored in `sales_order_item` table. The information regarding usages are stored in `product_options` column which is JSON encoded so we can't check or load only those orders which have custom option `usage_category` .

Reply:
Yes I can see how that can be difficult. It could be possible to use a regex to detect for the presence of usage_category but that may be tricky and cause other problems. We do need to do something here. Let's add a limit of 20 so it will just fetch the last twenty orders and no more. This will just be a best effort to get recent usages.